django_mongodb/compiler.py

```diff
@@ -591,42 +603,44 @@ def execute_sql(self, result_type):
                         f"{field.__class__.__name__}."
                     )
             prepared = field.get_db_prep_save(value, connection=self.connection)
-            values.append((field, prepared))
+            if hasattr(value, "as_mql"):
+                prepared = prepared.as_mql(self, self.connection)
+            values.append((field.column, prepared))
+        try:
+            criteria = self.build_query().mongo_query
+        except EmptyResultSet:
+            return 0
         is_empty = not bool(values)
-        rows = 0 if is_empty else self.update(values)
+        if is_empty:
+            rows = 0
+        else:
+            base_pipeline = [
+                {"$match": criteria},
+                {"$set": dict(values)},
+            ]
+            count_pipeline = [*base_pipeline, {"$count": "count"}]
+            pipeline = [
+                *base_pipeline,
+                {
+                    "$merge": {
+                        "into": self.collection_name,
+                        "whenMatched": "replace",
+                        "whenNotMatched": "discard",
+                    }
+                },
+            ]
+            with self.connection.connection.start_session() as session, session.start_transaction():
+                result = next(self.collection.aggregate(count_pipeline), {"count": 0})
+                self.collection.aggregate(pipeline)
```

Comment:
what has been the output when you merge:
`[*base_pipeline, {$merge..}, {$count...}]` Does this not work whatsoever?